### ISSUE TITLE
Can't specify which MetaMask Account (public address) to sign the message

### ISSUE THREAD

**Issue by eliasfeijo** [ChainSafe/web3.unity#134]
I'm using ChainSafe to create a Login with MetaMask functionality for a Desktop game.

The problem is: the method `Web3Wallet.Sign(message)` always opens the user's first MetaMask Account (public address) registered. The user can't switch to another Account in the dialog, and I can't specify the Account in the `Sign` method. In my application, the user inputs their Account public address in an Input Field. If the user wants to sign the message with any other account that is not the first, he can't. The MetaMask Sign message dialog automatically picks the first account for signing the message, and the user can't choose another public address to sign the message.

Is there a way to specify which MetaMask Account that will sign the message?

I'm guessing that this is also an issue with the `SendTransaction` method, if I can't specify the `from` wallet then my game will have problems.

**To Reproduce**
`String signature = await Web3Wallet.Sign("My Message");`

**Expected behavior**
The user should be able to select which MetaMask Account he will use to sign the message. Or the application should be able to specify which public address, from the user's wallet, should be used to sign the message.

**Screenshots**
The selected MetaMask Account was the second one, in the chrome extension, but the dialog always open with the first account and I can't change it. As you can see, the account name is grayed out, and I can't interact with it:
![MetaMaskLogin](https://user-images.githubusercontent.com/844450/164909980-f1496fd2-1798-4bbf-be5e-ab7dba262606.png)

As you can see below, MetaMask asks if it is the correct account, because the currently selected account (in the chrome extension) is different than the one being used to sign the message:
![MetaMaskLoginMessage](https://user-images.githubusercontent.com/844450/164913266-199cb8fd-987f-4928-b2c7-1a60d4a5e892.png)

**Desktop:**
 - OS: Ubuntu 20.04.4 LTS
 - Browser: Chrome

**Comment by eliasfeijo** [comment]
I've searched the internet for similar problems, and I've found this related post: https://forum.moralis.io/t/authenticate-with-metamask-always-use-the-first-account-and-not-the-selected-one/5409/18

I think the problem here is happening because I had already connected with my first account (Account 1) with the application, using ChainSafe. And after this first connection I can't connect any other accounts of my wallet, it always picks up the first account and it doesn't give me the option to switch.

Is this really is a bug on ChainSafe side, or am I missing something? I'm not very familiar with web3, so I don't know if it is possible for the application to select the MetaMask Account that will be used to sign the message. In the future I will also need to specify the `from` account of the `SendTransaction` method, because I'm storing the user's public address in a database, and the transaction must be sent from the correct address. Is this possible using this library?

**Comment by KBryan** [comment]
See if using the WebGLSignOut Prefab helps.
![WebGLSignOut](https://user-images.githubusercontent.com/681817/164976737-befab62c-56b6-47a5-b562-24bb343b9eda.png)

**Comment by eliasfeijo** [comment]
@KBryan I'm having trouble building for WebGL platform, so I can't test this prefab right now (I'm on Linux, Ubuntu 20.04).

But I've checked the prefab, and it only has a Button that calls `WebGLSignOut.OnSignOut()`, and this script is just doing this:

``` c#
public void OnSignOut()
{
    // Clear Account
    PlayerPrefs.SetString("Account", "0x0000000000000000000000000000000000000001");
    // go to login scene
    SceneManager.LoadScene(0);
}
```

Obviously it isn't invoking MetaMask for signing out, so it doesn't help me at all. I don't know why you've suggested this, but it's  not a solution. If this problem can't be fixed, just say so.

I might have to implement my own solution, instead of using this library, but it would be frustrating. I'm still hoping that this bug can easily be fixed by the maintainers.

**Comment by eliasfeijo** [comment]
I've managed to disconnect my MetaMask wallet by manually acessing the URL https://chainsafe.github.io/game-web3wallet/ and clicking on "Disconnect" inside the chrome's MetaMask extension. I've found this URL inside `Web3Wallet.cs` script.

It should be possible to have an action that invokes the disconnect wallet panel (for example: `?action=disconnect` query string in the URL, just like the other existing actions). If I could see the code behind this site, I'd gladly create a PR to fix this issue, but it seems that it's not open sourced yet.

Fow now, I can only hope for a maintainer to fix this issue for me. It would be even better if they could add the features that I've described [here](https://github.com/ChainSafe/web3.unity/issues/134#issue-1213328143) (eg. select a different wallet to connect to the game), but I don't know if those are possible using the current MetaMask API.

**Comment by eliasfeijo** [comment]
I've found that there is a way to disconnect an account using the MetaMask extension API: https://github.com/MetaMask/metamask-extension/issues/8990#issuecomment-980489771.

Since there was no response from the maintainers, I guess that fixing this issue is not in the project's roadmap right now. So I'll be working on my own dApp solution, and try to integrate it with Unity (probably by using this lib and just editing the dApp URL in the scripts).

**Comment by KBryan** [comment]
Thanks for posting this. Will add this to the backlog. There are many things we are working on and they are all important. If you would like to contribute to things you find useful. It's greatly appreciated.

**Comment by eliasfeijo** [comment]
Thanks, @KBryan! As I've commented previously, I'd gladly contribute if the dApp was open source (and I certainly would have forked the repo at this point). But, currently, I have specific use cases in my game that the library itself probably doesn't need to meet user expectations. So, I think I'm not really the best person to contribute right now. But if you need help with anything, PM me on discord, I'm @elf_qwerty in the ChainSafe server.

Here are some features that I'm planning to implement on my dApp, that currently don't exist in this library:

- Specify which wallet to be signing a message.
- Specify which wallet to be performing a transaction (`from` field of the MetaMask API)
- Display a button to disconnect a wallet from the dApp (this one isn't really required if the ones above work as I expect)

**Comment by sneakzttv** [comment]
in web3gl.jslib add
![image](https://user-images.githubusercontent.com/57473220/178848924-5c1d8e52-1fe4-4e66-96a9-aface9c0a29c.png)
in weblogin.cs do this and assign it to a disconnect button
![image](https://user-images.githubusercontent.com/57473220/178848942-7eee702d-d3b1-4079-be70-5ec01dcb0209.png)
and that'l trigger an account swap in metamask and you can do what you need from there

**Comment by eliasfeijo** [comment]
> in web3gl.jslib add ![image](https://user-images.githubusercontent.com/57473220/178848924-5c1d8e52-1fe4-4e66-96a9-aface9c0a29c.png) in weblogin.cs do this and assign it to a disconnect button ![image](https://user-images.githubusercontent.com/57473220/178848942-7eee702d-d3b1-4079-be70-5ec01dcb0209.png) and that'l trigger an account swap in metamask and you can do what you need from there

This doesn't help me, as I'm building for desktop and not for WebGL platform.

I've just implemented my own dApp solution instead of using the one provided by this library, which was really easy to create. If anyone is wondering about it, it's really simple. Then I added the feature of disconnecting the current selected account (which is kind of tricky, I had to search through [issues on MetaMask extension repository](https://github.com/MetaMask/metamask-extension/issues/8990#issuecomment-980489771) for a way to do it), and the other feature I mentioned, which is to allow a "from" address to be passed to the dApp, to be able to select which account should perform the transaction.
